Improve the first 40% of Documentation/CRDs/Block-Storage/ceph-block-pool-crd.md (#15809)
rook/rook: Documentation/CRDs/Block-Storage/ceph-block-pool-crd.md

title: CephBlockPool CRD
---

Rook enables creation and customization of Ceph block storage (RBD) pools through custom
resource definitions (CRDs). Settings and options are detailed below.

## Examples

### Replicated RBD Pool

For optimal performance, while also adding redundancy, this example configures
a Ceph RBD pool with three full copies of user data, each on three different nodes.

!!! note

>     This sample requires at least one OSD on each of at least three nodes.

Reviewer: IMO, the original reads more clearly than the rewrite here. I can imagine someone who doesn't have English as a native language getting confused here.

Author: Upstream Ceph follows the conventional English practice of spelling out small numbers.  This aligns with that.

  deviceClass: hdd
```

### Erasure Coded RBD Pool

This example specifies a pool with an EC 4+2 data protection scheme.  For a
given amount of raw storage, this pool yields double the usable space compared
to the replicated `size 3` example above.  Redundancy is provided by
specifying an [erasure coding](#erasure-coding) profile with `codingChunks: 2`,
akin to a conventional RAID6 volume but with much greater flexibility.  Data
remains available if any one OSD is down and is preserved if any two are lost.

> The downside is that write performance will be significantly lower compared to

Reviewer 2: With perf improvements coming in tentacle, how about we remove "significantly"? I've heard users say they see a 10-20% perf cost generally for EC, so if that's true, it doesn't seem like it's significant anyway. I like to avoid alarming wording where possible. :) 
```suggestion
The downside is that write performance will be lower compared to
```

Author: Re Tentacle, the scope of "fast EC" is not yet entirely clear to me, but based on a design document to which I was recently referred, I *think* those improvements are only limited to [over]writing partial stripes, i.e. RADOS objects whose size is` <= (replication * min_alloc_size)`

AIUI reads / writes of full RADOS object stripes are not affected, though I seek deeper understanding.  I also suspect at this time that these changes will not mitigate the padding space amp for very small RADOS objects, so I want to be careful to document accurately.

Reviewer 2: Yes, definitely a lot of factors to consider. Since there are so many factors from the hardware to configuration to the workload, our approach has been to leave it as an exercise for the reader to measure in their own clusters instead of trying to document.

  deviceClass: hdd
```

### Erasure Coded RBD Pool

This example specifies a pool with an EC 4+2 data protection scheme.  For a
given amount of raw storage, this pool yields double the usable space compared

> to the replicated ``size 3`` example above.  Redundancy is provided by

Reviewer: In the interest of making sure our website renderer doesn't do something unexpected, let's make sure to use single-grave for inline code.

```suggestion
to the replicated `size 3` example above.  Redundancy is provided by
```

> specifying an [erasure coding](#erasure-coding) profile with ``codingChunks: 2``,

Reviewer: ```suggestion
specifying an [erasure coding](#erasure-coding) profile with `codingChunks: 2`,
```

akin to a conventional RAID6 volume but with much greater flexibility.  Data

> remains available if any one node is down and is preserved if any two are lost.

Reviewer 2: Since the example sets `failureDomain: osd`. (edit: ignore this suggestion if we change the example to host failure domain)
```suggestion
remains available if any one OSD is down and is preserved if any two are lost.
```

Author: Ack.  Honestly the concept of OSD failure domain is so alien to me that I struggled a bit with this section.  Pushing a change.

> The downside is that client write performance will be roughly half that of the

Author: The subops are issued in parallel, there will be some skew in their completion, but mainly I'm thinking here of people using HDDs (or wacky VmWaRe LUNs etc) that are largely IOPS-bound.  4,2 vs R3 means twice the ops, modulo potentially fewer of them due to the sharding, it's a bit complex.  I'll remove the quantification.  For sure the impact on recovery and scrubs remains because of the gridlock / reservations.

above replicated pool, which notably means longer recovery (and thus increased
risk) when a server or OSD drive fails.

!!! note
    This example requires *at least* six OSDs*.

The OSDs can be located on a single Ceph node or spread across multiple nodes,
because the [`failureDomain`](ceph-block-pool-crd.md#spec) is set to `osd`.
The `erasureCoded` chunk settings require at least six OSDs of the specified
deviceClass: four `dataChunks` + two `codingChunks`).  We provide this example
of selecting the `osd` failure domain to demonstrate the idea, but for valuable
production data this is very risky: a single host going unavailable will mean
that at least some data is unavailable, and the complete loss of a single host's

> OSDs -- or *any* three OSDs -- will mean that some or all data is irretrievably lost.

Author: Sure.  I didn't want to take that much latitude as this is your turf, unlike upstream Ceph docs where I run rampant with wild abandon.  Especially as Rook's deployment domain may AFAIK include legitimate(ish) OSD failure domain clusters, but I'm onboard with not even telling people about OSD failure domain clusters.
Note that I changed this from 2,1 because nobody EVER should be doing 2,1 outside of a sandbox on their laptop and giving them an example just tempts fate.

> OSDs -- or *any* three OSDs -- will mean that some or all data is irretrievably lost.

Reviewer: Ditto the italics/bold question for "any". To help ensure clarity of intent, could you use either `_italic_` or `**bold**`

```suggestion
OSDs -- or _any_ three OSDs -- will mean that some or all data is irretrievably lost.
```

to full use the aggregate raw capacity when the failure domains are not
all equal in CRUSH weight.

A very small cluster that cannot provision at least six nodes, at least at first, may
choose to select an EC profile with `dataChunks` set to `3` or `2`, to accommodate
minima of five and four hosts, respectively, albeit with lower usable:raw capacity
ratios. A table showing the usable to raw space yields of various erasure coding profiles
can be viewed [here](https://docs.ceph.com/en/latest/rados/operations/erasure-code/#erasure-coded-pool-overhead).

Performance-sensitive applications typically will not use erasure coding due to
the performance overhead of distributing and gathering the multiple chunks (shards)
across the cluster.

### Mirroring

RADOS Block Device (RBD) mirroring is a process of asynchronous replication of
Ceph block device images (or entire pools) between two or more Ceph clusters.
Mirroring ensures point-in-time consistent data replication, including reads
and writes, volume (image) resizing, snapshots, clones, and flattening.
RBD mirroring is useful when planning for Disaster Recovery.
Mirroring is feasible for clusters that are geographically distributed

> with high network latency that precludes a `strtech mode` cluster or

Reviewer 2: ```suggestion
with high network latency that precludes a `strech mode` cluster or
```

Author: sigh, thought I fixed this.  See below.

spec:
  failureDomain: osd
  erasureCoded:
    dataChunks: 4
    codingChunks: 2
  deviceClass: hdd
```

Various combinations of `dataChunks` and `codingChunks` values are possible, with
concomitant tradeoffs.  Note that for valuable production data a `codingChunks`
value of 1 is *extremely* risky and is not advised.  Pools holding production data
should be created with a `codingChunks` value of at least 2.  Larger values of
`codingChunks` allow data to remain available across additional nodes being
down and survive the complete loss of additional OSDs or nodes, at the expense
of reduced usable capacity and write performance.

Clusters with valuable production data should comprise at least three nodes when using
replicated pools and at least four when using erasure coding.  Specifically, an
erasure-coded pool should specify `failureDomain: host` and the cluster should comprise at
least `dataChunks+1` hosts.  There are certain operational flexibility advantages to

> provisioning at least `dataChunks + codingChunks +1` hosts.

Author: This can be subtle: EC requires K+M failure domains at a minimum.  When there are only K+M failure domains, each shard is placed on every failure domain. Consider what happens if you have, say, 4+2 with six hosts, 5 hosts with CRUSH weight 100 and 1 host with CRUSH weight 200.  That extra 100 TB of CRUSH capacity can't ever be used, at lest not by the EC (and thus presumably dominant) pool.  If you have seven hosts, 6 @ 100 and 1 @ 200, then CRUSH can utilize more PGs on the larger failure domain than the others.  There are still combinatorial limits, having  6 @ 100 and 1 @ 1000 won't let all of that extra 900T capacity be utilized, but imbalances that large are much more rare.

This happens with replication too.  Consider an R3 pool in a cluster with domain weights 100, 100, 200.  The extra 100T of capacity on the third failure domain can't get used. 

ISTR having added a long explanation of this dynamic upstream, I didn't want to balloon the concise Rook docs by duplicating it all here.

I recently helped someone with a cluster of 10x nodes, each with 20x 20T HDD OSDs.  And dominant 8+2 pools. 
 They found two additional hosts to add in which improved recovery/backfill and frees them from losing 160T of raw capacity when a single drive fails.  A couple of their hosts were already down a couple of drives, so even if we hadn't added more OSDs at the same time, redistributing the existing OSDs would have resulted in more usable capacity.

> provisioning at least `dataChunks + codingChunks +1` hosts.

Reviewer 2: Ok, of course if the nodes are not balanced the additional host would allow more capacity to be used. Sounds good to keep the doc simple since that can be quite a complex topic to address. Perhaps a simple mention of imbalance?
```suggestion
provisioning at least `dataChunks + codingChunks +1` hosts, particularly if the OSDs are different sizes across hosts.
```

external-cluster-console # rbd mirror pool peer bootstrap import <token file path>
```

> See the official ``rbd-mirror`` documentation on [how to add a bootstrap peer](https://docs.ceph.com/docs/master/rbd/rbd-mirroring/#bootstrap-peers).

Reviewer: ```suggestion
See the official `rbd-mirror` documentation on [how to add a bootstrap peer](https://docs.ceph.com/docs/master/rbd/rbd-mirroring/#bootstrap-peers).
```

* `deviceClass`: Configure the CRUSH rule for this pool to distribute data only on OSDs of the specified device class. If left empty or unspecified, the pool will use the cluster's default CRUSH root, which usually distributes data over all OSDs, regardless of their class. If `deviceClass` is specified on any pool, ensure that it is added to *every* pool in the cluster, otherwise Ceph will warn about pools with overlapping roots. Additionally, the PG autoscaler and the Ceph Balancer may be confounded.  Be careful to examine the `.mgr` pool's CRUSH rule too.
* `crushRoot`: The root in the CRUSH topology to be used by the pool. If left empty or unspecified, the default root will be used. Creating a custom CRUSH root for OSDs currently requires the Rook toolbox to run the Ceph tools described [here](http://docs.ceph.com/docs/master/rados/operations/crush-map/#modifying-the-crush-map).
* `enableCrushUpdates`: Enables Rook to update the pool's CRUSH rule using Pool Spec. Can cause data remapping if the CRUSH rule is changed, Defaults to `false`.

> * `enableRBDStats`: Enables collecting RBD per-image IO statistics by enabling dynamic OSD performance counters. Defaults to `false`. For more info see the [ceph documentation](https://docs.ceph.com/docs/master/mgr/prometheus/#rbd-io-statistics). Note that this will be much more performance if the `object-map` and `fast-diff` RBD feature flags are present on RBD volumes.

Reviewer 2: ```suggestion
* `enableRBDStats`: Enables collecting RBD per-image IO statistics by enabling dynamic OSD performance counters. Defaults to `false`. For more info see the [ceph documentation](https://docs.ceph.com/docs/master/mgr/prometheus/#rbd-io-statistics). Note that this will be much more performant if the `object-map` and `fast-diff` RBD feature flags are present on RBD volumes.
```

Author: Clearly I picked the wrong week to give up amphetamines.

> * `name`: The name a Rook-deployed Ceph pool is based on the `metadata.name` of the CephBlockPool CR. Some built-in Ceph pools

Reviewer 2: Rook is implicit, so how about simplifying to this?
```suggestion
* `name`: The name of a Ceph pool is based on the `metadata.name` of the CephBlockPool CR. Some built-in Ceph pools
```

### Erasure Coding

[Erasure coding](http://docs.ceph.com/docs/master/rados/operations/erasure-code/) allows you to keep your data safe while reducing the storage overhead. Instead of creating multiple full replicas of the data,
erasure coding (EC) divides the original data into chunks of equal size, then generates additional chunks of that size for redundancy.  This is akin to RAID5 or RAID6 but much more flexible.

> For example, if you have a RADOS object of size 2MB, the simplest erasure coding with two data chunks would divide the object into two chunks of size 1MB each (data chunks). One more chunk (coding chunk) of size 1MB will be generated. In total, 3MB will be stored in the cluster. The object will be able to suffer the loss of any one of the chunks and still be able to reconstruct the original object.

Reviewer 2: How can we best phrase this in terms of user data? Users don't have RADOS objects, just files. What about mentioning files instead for clarity?
```suggestion
For example, if you have a file of size 2MB, the simplest erasure coding with two data chunks would divide the file into two chunks of size 1MB each (data chunks). One more chunk (coding chunk) of size 1MB will be generated. In total, 3MB will be stored in the cluster. The file will be able to suffer the loss of any one of the chunks and still be able to reconstruct the original file.
```

Author: Strictly speaking Ceph doesn't see files so I didn't want to confuse CephFS, RBD, and RGW.  I put in language using an RGW object as an example instead, does that work for you?  I also recast this in terms of a 4+2 profile given the below blessing to retcon 2+1.

Reviewer 2: RGW object sounds good, in addition to 4+2!

The number of data and coding chunks you choose will depend on your requirements for availability and data loss protection and how much storage overhead is acceptable.
Here are some examples to illustrate how the number of chunks affects the storage and loss toleration.  As we discuss above, M=1 is not recommended for irreplaceable data.
In most cases a value of K greater than 6 or 8 results in linear impact to write and recovery performance and demands additional failure domains, while the incremental
increase of the usable to raw capacities exhibits diminishing returns.

| Data chunks (k) | Coding chunks (m) | Total storage | Losses Tolerated | OSDs required |

> | --------------- | ----------------- | ------------- | ---------------- | ------------- |

Reviewer 2: While you're at it, want to remove the 2+1 row from this table since it's really not recommended?

Author: Love to, had left it out of deference.  Pushing.
